### ISSUE TITLE
OpenCore UEFI should support custom resolution

### ISSUE THREAD

**Issue by he110-c0de** [acidanthera/bugtracker#2464]
According to the log,  the resolution of UEFI screen is determined by OpenCore itself. BIOS framebuffer resolutions are just used for compassion. In fact some motherboard bios can set custom resolution.

So opencore should support custom resolutions which are not in the bios resolution list.I want this feature as soon as possible.

 07:986 00:391 OCC: Installing GOP (Unsupported) on ConsoleOutHandle...
08:015 00:028 OC: Requested resolution is 1920x1280@0 (max: 0, force: 1) from 1920x1280
08:043 00:028 OCC: Missing OcForceResolution protocol - Not Found
08:071 00:028 OCC: Requesting 1920x1280@0 (max: 0) resolution, curr 1, total 7
08:105 00:033 OCC: Current FB at 0xE0000000 (0x12C000), format 1, res 640x480 scan 640
08:133 00:028 OCC: Mode 0 - 1920x1080:1
08:162 00:028 OCC: Mode 1 - 640x480:1
08:187 00:025 OCC: Mode 2 - 800x600:1
08:216 00:028 OCC: Mode 3 - 1024x768:1
08:245 00:028 OCC: Mode 4 - 1280x1024:1
08:274 00:028 OCC: Mode 5 - 1400x1050:1
08:303 00:029 OCC: Mode 6 - 1280x960:1
**08:332 00:028 OCC: No compatible mode for 1920x1280@0 (max: 0) resolution**
08:361 00:029 OC: Changed resolution to 1920x1280@0 (max: 0, force: 1) from 1920x1280 - Not Found
09:810 01:449 OCC: MTRR supported, PAT supported
09:840 00:029 OCC: PAT 0x0007040600070406 (default)
09:870 00:029 OCC: 0xE0000000 MTRR 0=UC PTE2 bits 0x00000000E0000083 PAT@0->6=WB - Success
09:900 00:030 OCC: Writing PAT 0x0107040600070406
09:930 00:030 OCC: PAT 0x0107040600070406 (not default)
09:961 00:030 OCC: 0xE0000000 MTRR 0=UC PTE2 bits 0x00000000E000109B PAT@7->1=WC - Success
09:988 00:027 OCC: Switching to direct GOP renderer...
10:024 00:035 OCC: Found EG2 support 10002
10:055 00:030 OCC: Got rotation 0 degrees from EG2
10:086 00:031 OCBLT: 0: shl:0 shr:0 mask:FF0000
10:117 00:031 OCBLT: 1: shl:0 shr:0 mask:FF00
10:148 00:031 OCBLT: 2: shl:0 shr:0 mask:FF
10:179 00:031 OCBLT: Bytes per pixel: 4
10:211 00:031 OCBLT: 0: shl:0 shr:0 mask:FF0000
10:239 00:027 OCBLT: 1: shl:0 shr:0 mask:FF00
10:266 00:027 OCBLT: 2: shl:0 shr:0 mask:FF
10:298 00:031 OCBLT: Bytes per pixel: 4
10:330 00:031 OC: Setting UIScale to 1 - Success
10:392 00:062 OCC: Using builtin text renderer with 1 scale
10:450 00:057 OCC: Install console control (6EBB1E38/77729100/0), current - Success
10:508 00:058 OCC: Setup ASCII Output - Success
10:539 00:030 OC: Requested console mode is 0x0 (max: 0) from 
10:569 00:030 OCAU: OcAudioInstallProtocols (0, 0)
10:596 00:026 OCAU: 4B228577-6274-4A48-82AE-0713A1171987 protocol - Success
10:624 00:028 OCAU: C32332DF-FC56-4FE1-9358-BA0D529B24CD protocol - Success
10:656 00:031 OCAU: F4CB0B78-243B-11E7-A524-B8E8562CBAFA protocol - Success
10:692 00:036 OCAU: 3224B169-EC34-46D2-B779-E1B1687F525F protocol - Not Found
10:724 00:031 OCAU: Matching PciRoot(0x0)/Pci(0x1B,0x0)/VenMsg(A9003FEB-D806-41DB-A491-5405FEEF46C3,00000000)...
12:155 01:431 OCAU: 1/2 PciRoot(0x0)/Pci(0x3,0x0)/Pci(0x0,0x1)/VenMsg(A9003FEB-D806-41DB-A491-5405FEEF46C3,00000000) (4 outputs) - Success
12:188 00:032 OCAU: 2/2 PciRoot(0x0)/Pci(0x1B,0x0)/VenMsg(A9003FEB-D806-41DB-A491-5405FEEF46C3,00000000) (6 outputs) - Success
12:226 00:037 OCAU: System raw gain 0x46, audio mute (for chime) 0 - Success
12:259 00:032 OCAU: System decibel gain -17 dB - Success
12:292 00:033 OCAU: Setting language to en-US:0

**Comment by vit9696** [comment]
TL;DR You are misinterpreting the output of OpenCore log.

* OpenCore lists the resolutions it can choose from. None of them contain 1920x1280, thus it is correct the resolution remains intact.
* OpenCore does support setting custom resolutions via OcForceResolution protocol (exclusive to OpenDuet, works on few systems). Your system does not provide this protocol.

**Comment by he110-c0de** [comment]
@ vit9696 
I have did some google. I found that Linux/BSD can change the framebuffer resolution at boot time. So I think OpenCore should use the similar solution, keep the same with Linux/BSD.

https://www.reddit.com/r/archlinux/comments/sva9im/help_with_framebuffer_resolution/
https://www.antixforum.com/forums/topic/setting-the-native-resolution-in-the-framebuffer-console/
https://stackoverflow.com/questions/34904763/linux-framebuffer-set-resolution-correctly
